applications/ShapeOptimizationApplication/custom_utilities/damping/direction_damping_utilities.h

```diff
@@ -0,0 +1,263 @@
+// ==============================================================================
+//  KratosShapeOptimizationApplication
+//
+//  License:         BSD License
+//                   license: ShapeOptimizationApplication/license.txt
+//
+//  Main authors:    Armin Geiser, https://github.com/armingeiser
+//
+// ==============================================================================
+
+#ifndef DIRECTION_DAMPING_UTILITIES_H
+#define DIRECTION_DAMPING_UTILITIES_H
+
+// ------------------------------------------------------------------------------
+// System includes
+// ------------------------------------------------------------------------------
+#include <iostream>
+#include <string>
+#include <algorithm>
+
+// ------------------------------------------------------------------------------
+// Project includes
+// ------------------------------------------------------------------------------
+#include "includes/define.h"
+#include "includes/model_part.h"
+#include "spatial_containers/spatial_containers.h"
+#include "utilities/builtin_timer.h"
+#include "utilities/parallel_utilities.h"
+#include "custom_utilities/filter_function.h"
+#include "shape_optimization_application.h"
+
+// ==============================================================================
+
+namespace Kratos
+{
+
+///@name Kratos Classes
+///@{
+
+/// Class for direction damping of the shape update.
+/** Class for damping a specidfied direction from the shape update (and
+ * according to chain rule of differentiation also from sensitvities.)
+ * Can be used to constrain certain nodes of the design surface to slide on a plane.
+
+*/
+
+class DirectionDampingUtilities
+{
+public:
+    ///@name Type Definitions
+    ///@{
+
+    // Type definitions for better reading later
+    typedef array_1d<double,3> array_3d;
+    typedef ModelPart::NodeType NodeType;
+    typedef ModelPart::NodeType::Pointer NodeTypePointer;
+    typedef std::vector<NodeTypePointer> NodeVector;
+    typedef std::vector<NodeTypePointer>::iterator NodeVectorIterator;
+    typedef std::vector<double> DoubleVector;
+    typedef std::vector<double>::iterator DoubleVectorIterator;
+
+    // Type definitions for tree-search
+    typedef Bucket< 3, NodeType, NodeVector, NodeTypePointer, NodeVectorIterator, DoubleVectorIterator > BucketType;
+    typedef Tree< KDTreePartition<BucketType> > KDTree;
+
+    /// Pointer definition of DirectionDampingUtilities
+    KRATOS_CLASS_POINTER_DEFINITION(DirectionDampingUtilities);
+
+    ///@}
+    ///@name Life Cycle
+    ///@{
+
+    /// Default constructor.
+    DirectionDampingUtilities( ModelPart& modelPartToDamp, Parameters DampingSettings, int MaxNeighborNodes )
+        : mrModelPartToDamp( modelPartToDamp ),
+          mDampingSettings( DampingSettings ),
+          mMaxNeighborNodes( MaxNeighborNodes )
```

Comment:
why not adding this to the `Parameters`?

Reply:
makes sense, thanks for the other comments as well!